src/ut/realmmanager_test.cpp

```diff
@@ -149,6 +172,395 @@ TEST_F(RealmmanagerTest, CreateDestroy)
+
+  EXPECT_EQ(0, realm_manager->_failed_peers.size());
+
+  // We fail to connect to peer1 and peer4 and successufully connect
+  // to the other two peers.
+  realm_manager->peer_connection_cb(false, "1.1.1.1", DIAMETER_REALM);
+  realm_manager->peer_connection_cb(false, "4.4.4.4", DIAMETER_REALM);
+  realm_manager->peer_connection_cb(true, "2.2.2.2", DIAMETER_REALM);
+  realm_manager->peer_connection_cb(true, "3.3.3.3", DIAMETER_REALM);
+ 
+  // We expect only peer1 and peer4 to be added to _failed_peers
+  EXPECT_NE(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer1));
+  EXPECT_NE(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer4));
+  EXPECT_EQ(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer2));
+  EXPECT_EQ(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer3));
+  
+  EXPECT_EQ(2, realm_manager->_failed_peers.size());
+
+  // The diameter resolver returns the targets again. We expect to try and connect to peer1
+  // and peer4.
+  EXPECT_CALL(*_mock_resolver, resolve(DIAMETER_REALM, DIAMETER_HOSTNAME, 2, _, _))
+    .WillOnce(DoAll(SetArgReferee<3>(targets), SetArgReferee<4>(15)));
+  EXPECT_CALL(*_mock_stack, add(_))
+    .Times(2)
+    .WillRepeatedly(Return(true));
+  EXPECT_CALL(*_mock_stack, peer_count(4, 2))
+    .Times(1);
+
+  realm_manager->manage_connections(ttl);
+
+  // We now fail to connect to peer1, peer2 and peer3 and successfully
+  // connect to peer4.
+  realm_manager->peer_connection_cb(false, "1.1.1.1", DIAMETER_REALM);
+  realm_manager->peer_connection_cb(false, "3.3.3.3", DIAMETER_REALM);
+  realm_manager->peer_connection_cb(false, "2.2.2.2", DIAMETER_REALM);
+  realm_manager->peer_connection_cb(true, "4.4.4.4", DIAMETER_REALM);
+
+  // We expect peer1 to stay failed, peer2 and peer3 to be added to _failed_peers
+  // and peer4 to be removed from _failed_peers.
+  EXPECT_NE(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer1));
+  EXPECT_NE(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer2));
+  EXPECT_NE(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer3));
+  EXPECT_EQ(realm_manager->_failed_peers.end(), realm_manager->_failed_peers.find(peer4));
+
+  EXPECT_EQ(3, realm_manager->_failed_peers.size());
+
+  // We tidy up by having the resolver return no peers so that the RealmManager
+  // tears down it's connections.
+  targets.clear();
+  EXPECT_CALL(*_mock_resolver, resolve(DIAMETER_REALM, DIAMETER_HOSTNAME, 2, _, _))
+    .WillOnce(DoAll(SetArgReferee<3>(targets), SetArgReferee<4>(15)));
+  EXPECT_CALL(*_mock_stack, remove(_))
+    .Times(1);
+  EXPECT_CALL(*_mock_stack, peer_count(0, 0))
+    .Times(1);
+
+  realm_manager->manage_connections(ttl);
+
+  delete realm_manager;
+}
+
+// Test that we vorrectly remove old failed peers that are no longer being
+// returned by the DNS.
+TEST_F(RealmmanagerTest, TestRemoveOldFailedPeers)
```

Comment:
You're calling a private function of the RealmManager directly in the UT. The production code isn't able to do this - it only compiles in the UTs as we pass the no-access-control flag in (https://github.com/Metaswitch/clearwater-build-infra/blob/6298331d8befd9f3ddfeeef04b3abb4862901b26/cpp.mk#L192).
We try not to do this unless there's no other way to test this code, as it makes the UT less like the production code.